applications/Unity.GrantManager/modules/Unity.Payments/src/Unity.Payments.Application/PaymentRequests/PaymentRequestAppService.cs

```diff
@@ -197,5 +200,38 @@ protected virtual async Task<decimal> GetPaymentThresholdAsync()
 
             return PaymentSharedConsts.DefaultThresholdAmount;
         }
+        public async Task<string> GeneratePaymentNumberAsync()
+        {
+            var currentYear = DateTime.UtcNow.Year;
+            var prefix = "UP-";
+            var yearPart = currentYear.ToString();
+
+            // Get the next sequence number and reset it if it's a new year
+            var sequenceNumber = await GetNextSequenceNumberAsync(currentYear);
+            var sequencePart = sequenceNumber.ToString("D6");
+
+            return $"{prefix}{yearPart}-{sequencePart}";
+        }
+
+        private async Task<int> GetNextSequenceNumberAsync(int currentYear)
```

Comment:
Lets get the next sequence number up front ? We have a collection of payment requests to create - get the last number once, then increment one after another in the code without having to re-read from the database for each one

Reply:
Updated the code to limit the database call